source/slang/slang-ir-link.cpp

```diff
@@ -1519,6 +1520,146 @@ static void diagnoseUnresolvedSymbols(TargetRequest* req, DiagnosticSink* sink,
+            // we manage the "next" ptr with a seperate list instance since we don't want to add uses
+            // of an unrelated Call to our elementAddr
+            IRUse* next = nullptr;
+            for (auto use = i->firstUse; use; use = next)
+            {
+                next = use->nextUse;
+                auto user = use->user;
+                
+                switch (user->getOp())
+                {
+                case kIROp_StructFieldLayoutAttr:
+                {
+                    // ignore the def
+                    break;
+                }
+                case kIROp_Call:
+                { 
+                    builder.setInsertBefore(user);
+                    auto fieldAddress = builder.emitFieldAddress(
+                        builder.getPtrType(_dataField->getFieldType()),
+                        instToSwitch,
+                        _dataField->getKey()
+                        );
+                    auto elementAddr = builder.emitElementAddress(
+                        builder.getPtrType(builder.getUIntType()),
+                        fieldAddress,
+                        builder.getIntValue(builder.getIntType(), currOffset/4));
+
+                    user->setOperand(1, elementAddr);
+                    auto funcTypeInst = (user->getOperand(0));
+                    auto funcType = funcTypeInst->getFullType();
+
+                    auto paramReplacment = builder.getInOutType(builder.getUIntType());
+                    funcType->getOperand(1)->replaceUsesWith(paramReplacment);
+                    builder.addForceInlineDecoration(funcTypeInst);
+
+                    break;
+                }
+                }
+            }
+            if (i->typeUse.usedValue->getOp() == kIROp_GLSLAtomicUintType)
+            {
+                i->removeAndDeallocate();
+            }
+        }
+    }
+}
+
+void tryGLSLTypeReplacmentProcessing(IRSpecContext* context, TargetProgram* targetProgram, IRModule* irModule)
+{
+    // this is an issolated boolean operation since this operation can be moved to different
+    // parts of Slang and may likley need to be moved at some point
+    if (targetProgram->getOptionSet().getBoolOption(CompilerOptionName::AllowGLSL))
+    {
+        /// do 1 pass and get all data needed to process GLSL inst's that need attention
+        Dictionary<int, List<IRInst*>> bindingToInstMapUnsorted;
+        for (auto inst : irModule->getGlobalInsts())
+        {
+            if (inst->typeUse.usedValue)
+            {
+                switch (inst->typeUse.usedValue->getOp())
+                {
+                case kIROp_GLSLAtomicUintType:
+                {
+                    // Atomic_uint are supported by GLSL through converting to a different type (compiler ext)
+                    // not supported by SPIR-V->VK, means we must convert the type ourselves, else SPIR-V + 
+                    // HLSL is going to be impossible to emit as a target
```

Comment:
Please explain to my question

Reply:
I actually do mean '->' as in translate/convert/compile. SPIR-V->VK is an exact meaning. SPIR-V has no issues with AtomicCounter and AtomicStorage, it supports both. The issue is Vulkan spec dictating that atomic_uint and equivalents are not allowed in Vulkan, meaning when any driver is compiling SPIR-V->driver the code will fail to compile (SPIR-V compilers will throw an error if you let it to and tell it VK is the target) if using an atomic_uint equivlent in SPIR-V (AtomicCounter/AtomicStorage) 

This IR should bring up emit targets from what I understand since it will directly impact the emitted code (emit is supposed to only emit, not mangle types or do whole code passes), we need this code if we want to have the ability to emit [HLSL|SPIRV|GLSL]->VK/D3D12